--- camlp4/odyl/Makefile.mp-opt	Wed Oct  3 07:50:08 2001
+++ camlp4/odyl/Makefile	Sun Apr 21 20:09:12 2002
@@ -50,7 +50,10 @@
 
 install:
 	-$(MKDIR) $(LIBDIR) $(BINDIR)
-	cp odyl.cmo odyl.cma $(LIBDIR)
+	cp odyl.cmo odyl.cma odyl_main.cmi $(LIBDIR)
 	cp odyl$(EXE) $(BINDIR)/.
+	if [ -f odyl.cmxa ]; then\
+		cp odyl.cmxa odyl.a $(LIBDIR)/;\
+	fi
 
 include .depend
--- camlp4/ocaml_src/meta/Makefile.mp-opt	Tue Oct  2 07:49:03 2001
+++ camlp4/ocaml_src/meta/Makefile	Sun Apr 21 20:57:59 2002
@@ -2,7 +2,7 @@
 
 include ../../config/Makefile
 
-INCLUDES=-I ../camlp4 -I ../../boot -I $(OTOP)/utils
+INCLUDES=-I ../camlp4 -I ../../boot -I $(OTOP)/utils -I $(OTOP)/parsing
 OCAMLCFLAGS=$(INCLUDES)
 OBJS=q_MLast.cmo pa_r.cmo pa_rp.cmo pa_extend.cmo pa_extend_m.cmo pa_ifdef.cmo pr_dump.cmo
 CAMLP4RM=pa_r.cmo pa_rp.cmo pr_dump.cmo
--- camlp4/camlp4/Makefile.mp-orig	Tue Oct  2 19:25:20 2001
+++ camlp4/camlp4/Makefile	Mon May  6 23:27:20 2002
@@ -66,5 +66,6 @@
 	cp mLast.mli quotation.mli pcaml.mli spretty.mli $(LIBDIR)/.
 	cp mLast.cmi quotation.cmi ast2pt.cmi pcaml.cmi spretty.cmi $(LIBDIR)/.
 	cp camlp4.cma $(LIBDIR)/.
+	if [ -f camlp4.cmxa ]; then cp camlp4.cmxa camlp4.a $(LIBDIR)/.; fi
 
 include .depend
--- camlp4/etc/Makefile.mp-opt	Fri Sep 28 08:11:43 2001
+++ camlp4/etc/Makefile	Tue May  7 11:10:09 2002
@@ -5,12 +5,13 @@
 INCLUDES=-I ../camlp4 -I ../boot
 OCAMLCFLAGS=$(INCLUDES)
 OBJS=q_phony.cmo pa_o.cmo pa_op.cmo pa_oop.cmo pa_ru.cmo pa_format.cmo pa_olabl.cmo pa_sml.cmo pa_lisp.cmo pa_extfun.cmo pa_fstream.cmo pr_r.cmo pr_rp.cmo pr_o.cmo pr_op.cmo pr_extend.cmo pr_extfun.cmo pr_null.cmo pr_depend.cmo
+OBJSX=$(OBJS:.cmo=.cmx)
 INTF=pa_o.cmi
 CAMLP4OM=pa_o.cmo pa_op.cmo ../meta/pr_dump.cmo
-CAMLP4OMX=pa_o.cmx pa_op.cmx ../meta/pr_dump.cmx
+CAMLP4OMX=$(CAMLP4OM:.cmo=.cmx)
 SHELL=/bin/sh
 COUT=$(OBJS) camlp4o$(EXE)
-COPT=camlp4o.opt
+COPT=$(OBJSX) camlp4o.opt
 
 all: $(COUT) mkcamlp4.sh
 opt: $(COPT)
@@ -53,7 +54,7 @@
 	cp $(INTF) $(LIBDIR)/.
 	cp lib.sml $(LIBDIR)/.
 	cp camlp4o$(EXE) $(BINDIR)/.
-	if test -f $(COPT); then cp $(COPT) $(BINDIR)/.; fi
+	if test -f camlp4o.opt; then cp camlp4o.opt $(BINDIR)/.; cp $(OBJSX) $(OBJSX:.cmx=.o) $(LIBDIR)/.; fi
 	cp mkcamlp4.sh $(BINDIR)/mkcamlp4
 	chmod a+x $(BINDIR)/mkcamlp4
 
--- camlp4/meta/Makefile.mp-opt	Fri Sep 28 21:08:35 2001
+++ camlp4/meta/Makefile	Tue May  7 11:11:10 2002
@@ -2,14 +2,15 @@
 
 include ../config/Makefile
 
-INCLUDES=-I ../camlp4 -I ../boot -I $(OTOP)/utils
+INCLUDES=-I ../camlp4 -I ../boot -I $(OTOP)/utils -I $(OTOP)/parsing
 OCAMLCFLAGS=$(INCLUDES)
 OBJS=q_MLast.cmo pa_r.cmo pa_rp.cmo pa_extend.cmo pa_extend_m.cmo pa_ifdef.cmo pr_dump.cmo
+OBJSX=$(OBJS:.cmo=.cmx)
 CAMLP4RM=pa_r.cmo pa_rp.cmo pr_dump.cmo
 CAMLP4RMX=pa_r.cmx pa_rp.cmx pr_dump.cmx
 SHELL=/bin/sh
 COUT=$(OBJS) camlp4r$(EXE)
-COPT=camlp4r.opt
+COPT=$(OBJSX) camlp4r.opt
 
 all: $(COUT)
 opt: $(COPT)
@@ -46,6 +47,6 @@
 	cp $(OBJS) $(LIBDIR)/.
 	cp pa_ifdef.cmi $(LIBDIR)/.
 	cp camlp4r$(EXE) $(BINDIR)/.
-	if test -f $(COPT); then cp $(COPT) $(BINDIR)/.; fi
+	if test -f camlp4r.opt; then cp camlp4r.opt $(BINDIR)/.; cp $(OBJSX) $(OBJSX:.cmx=.o) $(LIBDIR)/.; fi
 
 include .depend
