--- camlp4/etc/pa_o.ml.orig	2004-08-18 04:17:37.000000000 -0700
+++ camlp4/etc/pa_o.ml	2004-09-06 15:00:59.000133744 -0700
@@ -554,7 +554,7 @@
           <:expr< let module $m$ = $mb$ in $e$ >>
       | "function"; OPT "|"; l = LIST1 match_case SEP "|" ->
           <:expr< fun [ $list:l$ ] >>
-      | "fun"; p = simple_patt; e = fun_def ->
+      | "fun"; p = patt LEVEL "simple"; e = fun_def ->
           <:expr< fun [$p$ -> $e$] >>
       | "match"; e = SELF; "with"; OPT "|"; l = LIST1 match_case SEP "|" ->
           <:expr< match $e$ with [ $list:l$ ] >>
@@ -715,7 +715,7 @@
   ;
   fun_binding:
     [ RIGHTA
-      [ p = simple_patt; e = SELF -> <:expr< fun $p$ -> $e$ >>
+      [ p = patt LEVEL "simple"; e = SELF -> <:expr< fun $p$ -> $e$ >>
       | "="; e = expr -> <:expr< $e$ >>
       | ":"; t = ctyp; "="; e = expr -> <:expr< ($e$ : $t$) >> ] ]
   ;
@@ -738,7 +738,7 @@
   ;
   fun_def:
     [ RIGHTA
-      [ p = simple_patt; e = SELF -> <:expr< fun $p$ -> $e$ >>
+      [ p = patt LEVEL "simple"; e = SELF -> <:expr< fun $p$ -> $e$ >>
       | "->"; e = expr -> <:expr< $e$ >> ] ]
   ;
   expr_ident:
@@ -789,11 +789,7 @@
     | LEFTA
       [ p1 = SELF; "."; p2 = SELF -> <:patt< $p1$ . $p2$ >> ]
     | "simple"
-      [ p = simple_patt -> p ] ]
-  ;
-
-  simple_patt:
-    [ [ s = LIDENT -> <:patt< $lid:s$ >>
+      [ s = LIDENT -> <:patt< $lid:s$ >>
       | s = UIDENT -> <:patt< $uid:s$ >>
       | s = INT -> <:patt< $int:s$ >>
       | s = INT32 -> MLast.PaInt32 loc s
@@ -988,7 +984,7 @@
     [ [ "="; ce = class_expr -> ce
       | ":"; ct = class_type; "="; ce = class_expr ->
           <:class_expr< ($ce$ : $ct$) >>
-      | p = simple_patt; cfb = SELF ->
+      | p = patt LEVEL "simple"; cfb = SELF ->
           <:class_expr< fun $p$ -> $cfb$ >> ] ]
   ;
   class_type_parameters:
@@ -996,11 +992,11 @@
       | "["; tpl = LIST1 type_parameter SEP ","; "]" -> (loc, tpl) ] ]
   ;
   class_fun_def:
-    [ [ p = simple_patt; "->"; ce = class_expr ->
+    [ [ p = patt LEVEL "simple"; "->"; ce = class_expr ->
           <:class_expr< fun $p$ -> $ce$ >>
       | p = labeled_patt; "->"; ce = class_expr ->
           <:class_expr< fun $p$ -> $ce$ >>
-      | p = simple_patt; cfd = SELF ->
+      | p = patt LEVEL "simple"; cfd = SELF ->
           <:class_expr< fun $p$ -> $cfd$ >>
       | p = labeled_patt; cfd = SELF ->
           <:class_expr< fun $p$ -> $cfd$ >> ] ]
@@ -1223,7 +1219,7 @@
     [ [ p = labeled_patt; e = SELF -> <:expr< fun $p$ -> $e$ >> ] ]
   ;
   labeled_patt:
-    [ [ i = LABEL; p = simple_patt ->
+    [ [ i = LABEL; p = patt LEVEL "simple" ->
            <:patt< ~ $i$ : $p$ >>
       | i = TILDEIDENT ->
            <:patt< ~ $i$ >>
