Index: Makefile
===================================================================
RCS file: /caml/ocaml/Makefile,v
retrieving revision 1.184
diff -u -u -r1.184 Makefile
--- Makefile	2003/07/03 15:13:21	1.184
+++ Makefile	2003/10/29 10:11:35
@@ -250,6 +250,10 @@
 	cp toplevel/toplevellib.cma $(LIBDIR)/toplevellib.cma
 	cp expunge $(LIBDIR)/expunge$(EXE)
 	cp typing/outcometree.cmi typing/outcometree.mli $(LIBDIR)
+	cp typing/typecore.cmi typing/typecore.mli $(LIBDIR)
+	cp parsing/location.cmi parsing/location.mli $(LIBDIR)
+	cp parsing/longident.cmi parsing/longident.mli $(LIBDIR)
+	cp parsing/parsetree.cmi parsing/parsetree.mli $(LIBDIR)
 	cp toplevel/topstart.cmo $(LIBDIR)
 	cp toplevel/toploop.cmi toplevel/topdirs.cmi toplevel/topmain.cmi $(LIBDIR)
 	cd tools; $(MAKE) install
